### PR TITLE
Initial jax+dask example.

### PR DESCRIPTION
**Pull request by asmith26**
This example is a learning exercise during the Scipy2020 Dask sprint to establish how dask might be used to parallelize [jax](https://github.com/google/jax)/[dm-haiku](https://github.com/deepmind/dm-haiku) deep learning model training and prediction. 

I've committed my notebook that is working end-to-end, and demonstrates a neural network for learning the [sine function](https://www.wolframalpha.com/input/?i=sin%28x%29). I appreciate this is quite a WIP, and I hope to improve it:
- [ ] Fix the training step (cell 10) to use dask/parallelization.
  - Any help on this would be greatly appreciated (I was originally thinking of using `dask.dataframe.map_partitions()`, but  I think this needs to return some dask dataframe/series object, while I want to return an updated `params, opt_state` tuple, so am now thinking dask.delayed will be best).
- [ ] Improve readability (e.g. move most of the code comments to dedicated markdown cells, add some background info, add clarity to images).

### CODE REVIEW

File: machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "@jax.jit\n",
+    "def predict(params: hk.Params, x: jnp.ndarray) -> jnp.ndarray:\n",
+    "    return net_transform.apply(params, x)\n",
+    "\n",
+    "\n",
+    "def dask_predict_wrapper(ddf: dd.DataFrame, params: hk.Params) -> jnp.ndarray:\n",
```
Comment (mrocklin) [inline machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb:299]:
```suggestion
    "def dask_predict_wrapper(df: pd.DataFrame, params: hk.Params) -> jnp.ndarray:\n",
```

Comment (mrocklin) [reply machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb:299]: I think that you meant for this to be a pandas dataframe.  It is used in a map_partitions call.  This function will be called on each of the pandas dataframes that make up the dask dataframe.

File: machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "for epoch_number in range(200):  # num_epochs\n",
+    "    for ddf_one_partition in ddf_train.partitions:  # for each batch\n",
+    "        df_one_partition = ddf_one_partition.compute()\n",
+    "        scaled_x = jnp.array(df_one_partition[[\"scaled_x\"]].values)\n",
+    "        y = jnp.array(df_one_partition[[\"y\"]].values)\n",
+    "        params, opt_state = update(params, opt_state, scaled_x, y)"
```
Comment (mrocklin) [inline machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb:393]: It might be worth taking a look at some of the functionality in dask-ml, which might do some of these things for you already if you're interested.

cc'ing @stsievert and @TomAugspurger

File: machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb (hunk 1 above)
Comment (asmith26) [reply machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb]: Good catch, thanks!

File: machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for epoch_number in range(200):  # num_epochs\n",
+    "    futures = []\n",
+    "    for ddf_one_partition in ddf_train.partitions:\n",
+    "        # Compute the gradients in parallel\n",
+    "        futures.append(client.submit(dask_compute_grads_one_partition_wrapper, ddf_one_partition, params))\n",
```
Comment (mrocklin) [inline machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb:483]: I recommend instead ... 

```python
from dask.distributed import futures_of
futures = futures_of(df.map_partitions(func, **params).persist())
```

File: machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+    "    \n",
+    "    for future in as_completed(futures):\n",
+    "        # Bring the gradients back to the client, and update the model with the optimizer on the client\n",
+    "        grads = future.result()\n",
```
Comment (mrocklin) [inline machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb:487]: See the `with_results=` keyword of as_completed to make this a bit faster.

File: machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+    "        updates, opt_state = optimizer.update(grads, opt_state)\n",
+    "        params = optix.apply_updates(params, updates)"
```
Comment (mrocklin) [inline machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb:489]: This is also the kind of thing for which Actors is probably a decent fit.

File: machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb (hunk 3 above)
Comment (asmith26) [reply machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb:483]: Thanks, I've tried this but `.map_partitions()` requires you to return either a Dask.DataFrame or Dask.Series (I think?). My function returns a set of gradients, `grads`, which is a Python dictionary (with more python dicts inside, i.e. a tree-like structure), so I don't think this will work in this case (please correct me if I am mistaken).

File: machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb (hunk 4 above)
Comment (asmith26) [reply machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb:487]: Thanks!

File: machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb (hunk 3 above)
Comment (TomAugspurger) [reply machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb:483]: You can probably work around that with `to_delayed()` instead of map_partitions. I can take a closer look later.

File: machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb (hunk 5 above)
Comment (asmith26) [reply machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb:489]: Yes, I've been trying to think how to perform training with shared parameters (and optimizer state) among workers via Actors, haven't quite got my head around how this might work yet.

Comment (mrocklin) [reply machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb:489]: This might be a start: https://docs.dask.org/en/latest/futures.html#example-parameter-server

Comment (stsievert) [reply machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb:489]: That example doesn't run, maybe a bad merge. I've put in a PR to correct that: https://github.com/dask/dask/pull/6449
